### PR TITLE
fix(extension): include CheckMetadataHash in the Substrate dApp signing payload

### LINKED ISSUE #4967: [Fix] Substrate dApp signPayload omits CheckMetadataHash, so every signature is rejected [Extension / Keysign]
# [Fix] Substrate dApp signPayload omits CheckMetadataHash, so every signature is rejected [Extension / Keysign]

## Problem

`constructPolkadotSigningPayload` (`core/ui/polkadot/dapp/constructSigningPayload.ts:42-51`) concatenates `method | era | nonce | tip | specVersion | transactionVersion | genesisHash | blockHash`. Both chains this route accepts carry the `CheckMetadataHash` signed extension, which contributes a `mode` byte to the extrinsic **extra** and an `Option<[u8;32]> metadataHash` to the **additional signed** data. We therefore sign a byte string 2 bytes shorter than the one the runtime reconstructs (34 bytes shorter when the dApp enables the metadata hash), and the node rejects the dApp's submission with `1010 Invalid Transaction: Transaction has a bad signature` (`BadProof`).

**Every Bittensor dApp signature fails.** This is total, not intermittent.

Live runtime metadata, both endpoints, today — the derived field sets are identical on both chains:

```
extra    : {"era","nonce","tip","mode":"u8"}
implicit : {"specVersion","transactionVersion","genesisHash","blockHash","metadataHash":"Option<[u8;32]>"}
```

`CheckMetadataHash` is the **last** signed extension in both tuples:

- Polkadot relay — `https://rpc.polkadot.io`, genesis `0x91b171bb158e2d3848fa23a9f1c25182fb8e20313b2c1eb49219da7a70ce90c3`, specName `polkadot`, spec 2005000 / txv 26 — 11 extensions, last = `CheckMetadataHash`.
- Bittensor finney — `https://entrypoint-finney.opentensor.ai`, genesis `0x2f0555cc76fc2840a25a6ea3b9637146806f1f44b090c175ffde2a7e5ab36c03`, specName `node-subtensor`, spec 467 / txv 1 — 13 extensions, last = `CheckMetadataHash`.

Those are exactly the two hashes hardcoded at `core/inpage-provider/popup/view/resolvers/sendTx/core/customTxData.ts:36-44`, so the route is wholly affected. Because the extension is last and everything after `ChargeTransactionPayment` is zero-sized on both chains, `mode` lands immediately after the tip and the `Option` immediately after the block hash — **both are append positions**.

The values already reach us. polkadot.js `toPayload()` unconditionally emits `mode: this.mode.toNumber()` and `metadataHash` (`node_modules/@polkadot/types/extrinsic/SignerPayload.js:114-116`), and transport is `JSON.stringify` → memo → `JSON.parse` (`clients/extension/src/inpage/providers/bittensor/index.ts:87` → `core/inpage-provider/popup/view/resolvers/sendTx/keysignPayload/build.ts:243` → `core/ui/mpc/keysign/action/mutations/useKeysignMutation.ts:221`). TypeScript types do not strip keys at runtime, so both fields are in the memo today — `PolkadotSignerPayload.ts:7-20` simply never declares them and the builder never reads them. `git grep -i "metadataHash\|CheckMetadataHash" -- . ':!*node_modules*'` returns nothing.

Our own native Bittensor signer already handles this correctly, which confirms the layout: `vultisig-android` `data/src/main/kotlin/com/vultisig/wallet/data/chains/helpers/BittensorHelper.kt:119` and `:133` write the byte into both `buildSignedExtra` (after the tip) and `buildAdditionalSigned` (after the block hash); its KDoc at `:20` says it bypasses wallet-core "because Bittensor requires the CheckMetadataHash signed extension".

**Scope note — lead with Bittensor, not Polkadot.** Since the Asset Hub Migration the relay holds 246,439 DOT against Asset Hub's 1,703,368,551 DOT (live `Balances::TotalIssuance`), and mainstream Polkadot dApp flows now use Asset Hub genesis, which `getSubstrateChain` rejects with `Unsupported Substrate genesis hash` before the builder runs. Relay impact is real but narrow (XcmPallet teleports, Proxy/Multisig/Utility, crowdloan refunds, coretime/OnDemand, residual-DOT transfers). The Bittensor half is unaffected by AHM and is where the traffic is.

## Expected Behavior

The bytes the vault signs match the bytes the runtime reconstructs, so the dApp's submission is accepted. Concretely:

```
method | era | compact(nonce) | compact(tip) | mode(1B)
  | u32le(specVersion) | u32le(transactionVersion) | genesisHash | blockHash | Option<[u8;32]>
```

with `Option` encoded as `0x00` when there is no metadata hash and `0x01 || <32 bytes>` when there is, and the >256-byte blake2b rule applied to the complete payload.

## Steps to Reproduce

User-facing: connect the extension to a Bittensor dApp (e.g. taostats.io), sign any transfer or stake. The vault signs, then the dApp reports `1010: Invalid Transaction: Transaction has a bad signature`.

Deterministic, no vault required — verify a real, already-accepted on-chain signature against both layouts:

1. Connect `@polkadot/api` to `wss://rpc.polkadot.io` and `wss://entrypoint-finney.opentensor.ai:443`.
2. Pull any signed extrinsic with `mode = 0` from a recent block; rebuild its signing payload twice — once with the current `constructPolkadotSigningPayload` concatenation, once with the `mode` byte after the tip and `Option::None` after the block hash.
3. Run `signatureVerify` against the extrinsic's own signer and signature.

Observed:

```
POLKADOT blk 33056594 onDemand.placeOrderKeepAlive  mode=0  len ext=101 canon=103
  EXTENSION layout verify: false | CANONICAL layout verify: true (sr25519)
BITTENSOR blk 9095861 subtensorModule.addStakeLimit mode=0  len ext=130 canon=132
  EXTENSION layout verify: false | CANONICAL layout verify: true (sr25519)
BITTENSOR blk 9095861 mevShield.submitEncrypted     mode=0  len ext=32  canon=32
  EXTENSION layout verify: false | CANONICAL layout verify: true (sr25519)
```

Signatures the real runtimes accepted verify only against the layout we do not build.

## Solution

Declare `mode?: number` and `metadataHash?: string | null` on `PolkadotSignerPayloadJSON` (`core/ui/polkadot/dapp/PolkadotSignerPayload.ts:7-20`), then in `constructSigningPayload.ts` insert the `mode` byte into the concatenation after `tip` and the SCALE-encoded `Option<[u8;32]>` after `blockHash`, before the length check. Both are append positions within their group because `CheckMetadataHash` is last in the signed-extension tuple on both accepted chains, so nothing else shifts.

**Do not hardcode `0x00` / `0x00`.** Sampling live Bittensor traffic, 51 of 351 signed extrinsics (~15%) carry `mode = 1`, i.e. a real `Some([u8;32])` implicit — a 34-byte gap, not 2. The `mode = 1` calls are exactly what a dApp requests: `balances.transferKeepAlive`, `subtensorModule.addStake` / `removeStake` / `addStakeLimit` / `removeStakeLimit`, `utility.batchAll`. A zero-hardcoded fix would still fail for one in seven real transactions. Read both values from the payload, defaulting to `0` / `None` when absent so existing behaviour is preserved for payloads that omit them.

Note `mode` arrives as a **number** and `metadataHash` as a hex string or `null` — do not pass `mode` through `parseInt(x, 16)` the way `nonce`, `specVersion` and `transactionVersion` are handled in the same function.

### Must Do
- [ ] Add `mode?: number` and `metadataHash?: string | null` to `PolkadotSignerPayloadJSON`.
- [ ] Append the `mode` byte after `tip` and the SCALE `Option<[u8;32]>` after `blockHash` in `constructPolkadotSigningPayload`, reading both from the payload (defaults `0` / `None`).
- [ ] Apply the existing >256-byte blake2b rule to the complete payload, after both insertions.
- [ ] Cover the builder with a unit test — it currently has none: assert the byte layout for `mode = 0` and for `mode = 1` with a 32-byte hash, and that a payload crossing 256 bytes only after the insertions is hashed.

### Must NOT Do
- Don't change unrelated code
- Don't refactor surrounding logic
- Don't touch the native Send path or `signRaw` — both are unaffected
- Don't hardcode the two bytes as zero

## Acceptance Criteria
- [ ] `yarn check:all` succeeds
- [ ] For a `mode = 0` payload, the builder's output equals the canonical payload and is 2 bytes longer than today's
- [ ] For a `mode = 1` payload with a metadata hash, the output carries `0x01 || <32 bytes>` after `blockHash` and is 34 bytes longer than today's
- [ ] A payload missing both fields still produces today's bytes plus `0x00` / `0x00`, not a crash
- [ ] Signing from a Bittensor dApp submits successfully instead of returning `1010 ... Transaction has a bad signature`

### PR DESCRIPTION
**Pull request by Ehsan-saradar**
## Description

Fixes #4967

### Problem

`constructPolkadotSigningPayload` builds the bytes a Substrate dApp `signPayload` request gets signed over as

```
method | era | compact(nonce) | compact(tip) | u32le(specVersion) | u32le(transactionVersion) | genesisHash | blockHash
```

Both genesis hashes `customTxData.ts` accepts — Polkadot relay and Bittensor finney — carry the `CheckMetadataHash` signed extension, and it is the **last** extension in both tuples (verified against live metadata on both RPCs). It contributes a `mode: u8` to the extrinsic extra and an `Option<[u8;32]> metadataHash` to the implicit data, so the runtime reconstructs

```
method | era | compact(nonce) | compact(tip) | u8(mode)
  | u32le(specVersion) | u32le(transactionVersion) | genesisHash | blockHash | Option<[u8;32]>
```

We signed a string 2 bytes shorter (34 when the dApp enables the metadata hash), and every dApp submission came back `1010: Invalid Transaction: Transaction has a bad signature`.

The values were already in the memo: polkadot.js `toPayload()` unconditionally emits `mode` (number) and `metadataHash` (hex or `null`), and the payload is `JSON.stringify`'d verbatim into the keysign memo. `PolkadotSignerPayloadJSON` simply never declared the fields and the builder never read them.

### Fix

- `PolkadotSignerPayloadJSON` declares `mode?: number` and `metadataHash?: string | null`.
- The builder appends `u8(mode)` after the tip and the SCALE `Option<[u8;32]>` after the block hash, defaulting to `0` / `None` when a payload omits them (what the runtime reconstructs for a dApp that did not opt in). The >256-byte blake2b rule applies to the complete payload.
- Since the payload comes from an untrusted dApp, three fail-fast guards: `mode` must be `0` or `1` (a `u8` would otherwise wrap silently — the runtime decodes nothing else); `mode` and `metadataHash` must agree, because the runtime derives the implicit as `mode == 1 ? Some(hash) : None`; and a supplied `metadataHash` (including an empty string) must be 32 bytes. All throw a local error instead of producing a signature the chain rejects.

Not touched: the native Send path, `signRaw`, and the two call sites in `useKeysignMutation.ts` / `FastKeysignServerStep.tsx`, which already pass the parsed memo through unchanged.

### Tests

New `constructSigningPayload.test.ts` (the builder had none):

- **Two real Bittensor extrinsics** from block 9101269, both accepted by the runtime, one 120-byte raw payload and one 401-byte hashed payload. The test rebuilds each from its signer-payload fields and checks the on-chain sr25519 signature with `signatureVerify` — it passes against the new layout and a negative case shows it fails against the old one.
- Exact byte layouts for `mode = 0` and `mode = 1 + hash`; the +2 / +34 length deltas; omitted fields equal `0` / `None`; a payload that crosses 256 bytes only because of the appended `Option` is hashed while one landing exactly on 256 is not; every rejection path (bad mode, mode/hash disagreement both ways, wrong-length and empty hash).

A real `mode = 1` signature is not included: the extrinsic only carries the `mode` byte, and the hash itself is derived from merkleized runtime metadata that needs tooling this repo does not have. That case is covered at the byte-layout level, and the ordering is the same `Option<[u8;32]>` codec polkadot.js applies.

### Verification

`yarn check` ✅ · `yarn test` ✅ 1660 passed (220 files) · extension `tsc --noEmit` (ES2021 lib) ✅

Not verified by me: an end-to-end sign from a Bittensor dApp. That's the last acceptance criterion on the issue and needs a vault connected to a live dApp.

## Which feature is affected?
- [ ] Create vault ( Secure / Fast)
- [ ] Sending
- [ ] Swap
- [x] New Chain / Chain related feature — Substrate dApp `signPayload` (Bittensor, Polkadot relay)

## Parity

- iOS: n/a — no Substrate dApp `signPayload` route; the native `BittensorHelper` already encodes `CheckMetadataHash`
- Android: n/a — same; native `BittensorHelper` already encodes it
- SDK: n/a — the dApp signer-payload builder lives only in this repo's `core/ui`; the SDK's Bittensor `buildExtrinsic` already encodes the extension for native sends

## Checklist

- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works

## Screenshots (if applicable):

n/a

## Agent Metadata (if applicable)
- **Agent**: Claude Code
- **Issue**: #4967
- **Confidence**: high — the layout is pinned by real on-chain signatures, not by reasoning about the metadata alone
- **Needs human review for**: a live sign from a Bittensor dApp (the one acceptance criterion not covered offline)

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

- **New Features**
  - Added support for Polkadot `CheckMetadataHash` signing fields, including signing modes and metadata hashes where applicable.
  - Signing payloads now validate supported modes and metadata hash formats.
  - Payloads exceeding 256 bytes continue to use hashing, while payloads exactly 256 bytes remain un-hashed.

- **Bug Fixes**
  - Improved compatibility of generated signing payloads with sr25519 signatures and metadata hash extensions.
  - Added validation for missing, invalid, or mismatched metadata hash values.

### CODE REVIEW

File: core/ui/polkadot/dapp/constructSigningPayload.ts [hunk 1]
```diff
@@ -21,6 +62,7 @@ export const constructPolkadotSigningPayload = (
   const era = hexToU8a(payload.era)
   const nonce = compactToU8a(parseInt(payload.nonce, 16))
   const tip = compactToU8a(payload.tip ? BigInt(payload.tip) : 0n)
+  const mode = encodeCheckMetadataHashMode(payload.mode)
```
[the remark this answers is not in the document]
Comment (Ehsan-saradar) [reply core/ui/polkadot/dapp/constructSigningPayload.ts]: Skipping this one on purpose. What the runtime reconstructs is determined by the chain, not by the dApp's list, and this route only accepts two hardcoded genesis hashes (`customTxData.ts`) — Polkadot relay and Bittensor — both of which declare `CheckMetadataHash` (checked against live metadata on both RPCs). Gating on a dApp-supplied `signedExtensions` could only ever make us sign bytes the chain rejects, never bytes it accepts, and it would throw on any payload that omits the field. If a chain without the extension is ever added to the accepted set, the builder is the right place to handle it then.

**Review by johnnyluo — APPROVED** [review]
LGTM
